File: src/aircraft/game/bomb/Bomb.java
package aircraft.game.bomb;

import java.awt.geom.Point2D;
import java.awt.Graphics;
import java.awt.image.BufferedImage;

import aircraft.game.Canvas;
import aircraft.game.ImageLoader;
import aircraft.game.Setting;
import aircraft.game.plane.Plane;

public class Bomb {
  protected final Canvas canvas;
  public BufferedImage image;

  // The current position.
  public Point2D.Double location = new Point2D.Double();
  protected double speed = 1; // The speed of the flying object.

  public Bomb(Canvas canvas, double x, double y) {
    // Load the bomb from the image directory.
    this.canvas = canvas;
    image = ImageLoader.readImg(Setting.images.get("Bomb"));
    if (image.getWidth()  >= canvas.getWidth() ||
        image.getHeight() >= canvas.getHeight())
      throw new RuntimeException("The size of image is invalid");
    
    location.x = x; location.y = y;
  }

  // The display method: as the name suggests, this method draw the image
  // onto the graphics user interface, so the argument is necessary.
  public void display(Graphics graphics) {
    // Default implementation: show image on the gui.
    graphics.drawImage(image, (int)location.x, (int)location.y, null);
  }

  // Check whether the bomb is displayed inside the window.
  public void boundaryCheck() {
    double xmin = location.x; // Left boundary.
    double xmax = location.x + image.getWidth(); // Right boundary.
    double ymin = location.y; // Bottom boundary.
    double ymax = location.y + image.getHeight(); // Top boundary.

    if (xmin < 0 || xmax > canvas.getWidth() ||
        ymin < 0 || ymax > canvas.getHeight())
      canvas.objects.get("trash").add(this);
  }

  // Each object should have a specific move policy.
  public void move() {
    location.y += speed;
  }

  public void effect(Plane plane) {
    if (plane.getHP() > 0) plane.clearHP();
  }

  // Return the camp of this object.
  public int camp() {
    return Setting.NEUTRAL;
  }
}

File: python/bomb.py
from point2d import Point2D
import imgloader
import setting

class Bomb:
    def __init__(self, canvas, x, y):
        """Initialize the bomb class.

        Keyword arguments:
        canvas -- the canvas for gui.
        x -- the x coordinate of location.
        y -- the y coordinate of location.
        """
        # Read image and check its size.
        self.canvas = canvas
        self.image = imgloader.load(setting.images["Bomb"])
        w, h = self.image.get_rect().size
        if w >= self.canvas.width or h >= self.canvas.height:
            raise ValueError("The size of image is invalid")

        self.location = Point2D(x, y)
        self.speed = setting.speed["Bomb"] # The speed of the flying object.

    def display(self, graphics):
        """Display image on the screen.
        """
        # Show image on the screen with respect to the specified location
        # and the image sources. Note that this method is not abstract.
        graphics.blit(self.image, self.location.cartesian())

    def boundary_check(self):
        """Check whether the bomb moves out of window.

        One bomb will be directly moved to trash and deleted in the next frame
        once it moves out of the current game window.
        """
        w, h = self.image.get_rect().size
        if self.location.x < 0 or self.location.x + w > self.canvas.width or \
           self.location.y < 0 or self.location.y + h > self.canvas.height:
            self.canvas.objects["trash"].add(self)

    def move(self):
        """Move the bomb to the next location at the next frame, according to
        the current speed and direction.
        """
        self.location.y += self.speed

    def effect(self, plane):
        """A bomb make the plane explode directly.
        """
        if plane.health > 0: plane.health = 0
    
    def camp(self):
        """Return the camp of this object.
        """
        return setting.NEUTRAL
